test/e2e/autoscale_hpa_test.go

```diff
@@ -0,0 +1,205 @@
+// +build e2e
+// +build hpa
+
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	vegeta "github.com/tsenart/vegeta/v12/lib"
+	"golang.org/x/sync/errgroup"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/pkg/test/spoof"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	resourcenames "knative.dev/serving/pkg/reconciler/revision/resources/names"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	v1test "knative.dev/serving/test/v1"
+)
+
+const (
+	cpuTarget             = 75
+	targetPods            = 5
+	scaleUpTimeout        = 3 * time.Minute
+	scaleToMinimumTimeout = 10 * time.Minute // 5 minutes is the default window for hpa to calculate if should scale down
+	minimumReplicas       = 1
+	minPods               = 1.0
+	maxPods               = 10.0
+	primeNum              = 500000
+)
+
+func TestHPAAutoscaleUpDownUp(t *testing.T) {
+	ctx := setupHPASvc(t, autoscaling.HPA, autoscaling.CPU, cpuTarget)
+	test.EnsureTearDown(t, ctx.Clients(), ctx.Names())
+	assertHPAAutoscaleUpToNumPods(ctx, minimumReplicas, targetPods, time.After(scaleUpTimeout), true /* quick */)
+	assertScaleDownToOne(ctx)
+	assertHPAAutoscaleUpToNumPods(ctx, minimumReplicas, targetPods, time.After(scaleUpTimeout), true /* quick */)
+}
+
+func setupHPASvc(t *testing.T, class, metric string, target int) *TestContext {
+	t.Helper()
+	clients := Setup(t)
+
+	t.Log("Creating a new Route and Configuration")
+	names := &test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   autoscaleTestImageName,
+	}
+	resources, err := v1test.CreateServiceReady(t, clients, names,
+		[]rtesting.ServiceOption{
+			rtesting.WithConfigAnnotations(map[string]string{
+				autoscaling.ClassAnnotationKey:    class,
+				autoscaling.MetricAnnotationKey:   metric,
+				autoscaling.TargetAnnotationKey:   strconv.Itoa(target),
+				autoscaling.MinScaleAnnotationKey: fmt.Sprintf("%d", minimumReplicas),
+				autoscaling.MaxScaleAnnotationKey: fmt.Sprintf("%d", int(maxPods)),
+			}), rtesting.WithResourceRequirements(corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("30m"),
+					corev1.ResourceMemory: resource.MustParse("20Mi"),
+				},
```

Comment:
Should we also set limits? We're going to consume quite a lot of resources without that, aren't we?

Reply:
Yes indeed but also we dont want some mem spike to kill the pods (I assumed we have big enough nodes) but anyway I can size the pods.

Reply:
We can limit just the CPU then, I think. Limiting concurrency as suggested below might do the trick as well.

Reply:
Ok I will experiment.